### ISSUE TITLE
misleading redefined warnings

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5314]
Migrated from [rt.perl.org#8915](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=8915) (status was 'resolved')

Searchable as RT8915$

**Comment by p5pRT** [comment]
### From stas@stason.org

#### Created by stas@hope.stason.org



Consider these three one liners&#8203;:

% perl \-we 'sub a \{\} \*\{"a"\} = sub \{\}'
Subroutine a redefined at \-e line 1\.

% perl \-we 'sub My&#8203;::a \{\} sub My&#8203;::a\{\}'   
Subroutine My&#8203;::a redefined at \-e line 1\.

% perl \-we 'sub My&#8203;::a \{\} \*\{"My&#8203;::a"\} = sub \{\}'
Subroutine a redefined at \-e line 1\.

The first two are fine\, the third one loses the package name in report\. Why the second 
one liner reports the whole things as it's used\, whereas the third one loses the 
package name?

This is a problem if functions from different packages get redefined \(e\.g\. via do\(\)  
or else\) and it would be nice if the third case will not lose the package name\.

Also I see no reason why the first one liner shouldn't warn with&#8203;:

&nbsp;  Subroutine main&#8203;::a redefined at \-e line 1\.

to be always consistent\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.7.3:

Configured by stas at Mon Apr  8 00:55:30 SGT 2002.

Summary of my perl5 (revision 5.0 version 7 subversion 3 patch 15780) configuration:
  Platform:
    osname=linux, osvers=2.4.17-16mdksmp, archname=i686-linux-thread-multi
    uname='linux hope.stason.org 2.4.17-16mdksmp #1 smp fri feb 8 14:27:48 cet 2002 i686 unknown '
    config_args='-des -Dprefix=/home/stas/perl/ithread -Dusethreads -Doptimize=-g -Duseshrplib -Dusedevel'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='2.96 20000731 (Mandrake Linux 8.2 2.96-0.74mdk)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.4.so, so=so, useshrplib=true, libperl=libperl.so
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/home/stas/perl/ithread/lib/5.7.3/i686-linux-thread-multi/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL15771


@INC for perl v5.7.3:
    /home/stas/perl/ithread/lib/5.7.3/i686-linux-thread-multi
    /home/stas/perl/ithread/lib/5.7.3
    /home/stas/perl/ithread/lib/site_perl/5.7.3/i686-linux-thread-multi
    /home/stas/perl/ithread/lib/site_perl/5.7.3
    /home/stas/perl/ithread/lib/site_perl
    .


Environment for perl v5.7.3:
    HOME=/home/stas
    LANG=en
    LANGUAGE=en_US:en
    LC_COLLATE=en_US
    LC_CTYPE=en_US
    LC_MESSAGES=en_US
    LC_MONETARY=en_US
    LC_NUMERIC=en_US
    LC_TIME=en_US
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr//bin:/bin:/usr/bin:.:/usr/local/bin:/home/stas/bin:/home/stas/bin:/usr/local/bin:/usr/X11R6/bin:/usr/java/j2re1.4.0/bin/:/home/stas/src/insure/bin.linux2
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

On 2002\.04\.07 19&#8203;:36 Stas Bekman wrote&#8203;:
> Consider these three one liners&#8203;:
> 
> % perl \-we 'sub a \{\} \*\{"a"\} = sub \{\}'
> Subroutine a redefined at \-e line 1\.
> 
> % perl \-we 'sub My&#8203;::a \{\} sub My&#8203;::a\{\}'   
> Subroutine My&#8203;::a redefined at \-e line 1\.
> 
> % perl \-we 'sub My&#8203;::a \{\} \*\{"My&#8203;::a"\} = sub \{\}'
> Subroutine a redefined at \-e line 1\.
> 
> The first two are fine\, the third one loses the package name in report\. Why the second 
> one liner reports the whole things as it's used\, whereas the third one loses the 
> package name?
> 
> This is a problem if functions from different packages get redefined \(e\.g\. via do\(\)  
> or else\) and it would be nice if the third case will not lose the package name\.
> 
> Also I see no reason why the first one liner shouldn't warn with&#8203;:
> 
>   Subroutine main&#8203;::a redefined at \-e line 1\.
> 
> to be always consistent\.

This should be fixed by change \#15789\.

**Comment by p5pRT** [comment]
### From @rgarcia

On 2002\.04\.07 22&#8203;:36 Rafael Garcia\-Suarez wrote&#8203;:
> On 2002\.04\.07 19&#8203;:36 Stas Bekman wrote&#8203;:
> > Consider these three one liners&#8203;:
> > 
> > % perl \-we 'sub a \{\} \*\{"a"\} = sub \{\}'
> > Subroutine a redefined at \-e line 1\.
> > 
> > % perl \-we 'sub My&#8203;::a \{\} sub My&#8203;::a\{\}'   
> > Subroutine My&#8203;::a redefined at \-e line 1\.
> > 
> > % perl \-we 'sub My&#8203;::a \{\} \*\{"My&#8203;::a"\} = sub \{\}'
> > Subroutine a redefined at \-e line 1\.
> > 
> > The first two are fine\, the third one loses the package name in report\. Why the second 
> > one liner reports the whole things as it's used\, whereas the third one loses the 
> > package name?
> > 
> > This is a problem if functions from different packages get redefined \(e\.g\. via do\(\)  
> > or else\) and it would be nice if the third case will not lose the package name\.
> > 
> > Also I see no reason why the first one liner shouldn't warn with&#8203;:
> > 
> >   Subroutine main&#8203;::a redefined at \-e line 1\.
> > 
> > to be always consistent\.
> 
> This should be fixed by change \#15789\.

D'oh\, here is the applied patch : this changes the warning message
for subroutines redefined via a typeglob assignment\.

<details><summary>Inline Patch</summary>

```diff
--- t/lib/warnings/op.orig	Mon Apr  1 21:00:09 2002
+++ t/lib/warnings/op	Sun Apr  7 22:26:47 2002
@@ -775,7 +775,7 @@
 sub fred () { 1 }
 *fred = sub () { 2 };
 EXPECT
-Constant subroutine fred redefined at - line 4.
+Constant subroutine main::fred redefined at - line 4.
 ########
 # op.c
 use warnings 'redefine' ;
--- t/lib/warnings/sv.orig	Mon Nov 19 00:01:48 2001
+++ t/lib/warnings/sv	Sun Apr  7 22:25:47 2002
@@ -262,7 +262,7 @@
 sub jim {} 
 *jim = \&joe ;
 EXPECT
-Subroutine fred redefined at - line 5.
+Subroutine main::fred redefined at - line 5.
 ########
 # sv.c
 use warnings 'printf' ;
--- sv.c.orig	Thu Mar 28 15:33:31 2002
+++ sv.c	Sun Apr  7 22:13:13 2002
@@ -3839,8 +3839,9 @@ Perl_sv_setsv_flags(pTHX_ SV *dstr, regi
  				{
  				    Perl_warner(aTHX_ packWARN(WARN_REDEFINE),
  					CvCONST(cv)
- 					? "Constant subroutine %s redefined"
- 					: "Subroutine %s redefined",
+ 					? "Constant subroutine %s::%s redefined"
+ 					: "Subroutine %s::%s redefined",
+					HvNAME(GvSTASH((GV*)dstr)),
  					GvENAME((GV*)dstr));
  				}
 			    }
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Rafael Garcia\-Suarez wrote&#8203;:
> On 2002\.04\.07 19&#8203;:36 Stas Bekman wrote&#8203;:
> 
>>Consider these three one liners&#8203;:
>>
>>% perl \-we 'sub a \{\} \*\{"a"\} = sub \{\}'
>>Subroutine a redefined at \-e line 1\.
>>
>>% perl \-we 'sub My&#8203;::a \{\} sub My&#8203;::a\{\}'   
>>Subroutine My&#8203;::a redefined at \-e line 1\.
>>
>>% perl \-we 'sub My&#8203;::a \{\} \*\{"My&#8203;::a"\} = sub \{\}'
>>Subroutine a redefined at \-e line 1\.
>>
>>The first two are fine\, the third one loses the package name in report\. Why the second 
>>one liner reports the whole things as it's used\, whereas the third one loses the 
>>package name?
>>
>>This is a problem if functions from different packages get redefined \(e\.g\. via do\(\)  
>>or else\) and it would be nice if the third case will not lose the package name\.
>>
>>Also I see no reason why the first one liner shouldn't warn with&#8203;:
>>
>>  Subroutine main&#8203;::a redefined at \-e line 1\.
>>
>>to be always consistent\.
> 
> 
> This should be fixed by change \#15789\.

Confirmed that it does the right thing now :\) Thanks Rafael\!

\-\- 


\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_
Stas Bekman            JAm\_pH \-\-\-\-\-\-> Just Another mod\_perl Hacker
http&#8203;://stason\.org/     mod\_perl Guide \-\-\-> http&#8203;://perl\.apache\.org
mailto&#8203;:stas@&#8203;stason\.org http&#8203;://use\.perl\.org http&#8203;://apacheweek\.com
http&#8203;://modperlbook\.org http&#8203;://apache\.org   http&#8203;://ticketmaster\.com

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

#### Created by stas@hope.stason.org



Consider these three one liners&#8203;:

% perl \-we 'sub a \{\} \*\{"a"\} = sub \{\}'
Subroutine a redefined at \-e line 1\.

% perl \-we 'sub My&#8203;::a \{\} sub My&#8203;::a\{\}'   
Subroutine My&#8203;::a redefined at \-e line 1\.

% perl \-we 'sub My&#8203;::a \{\} \*\{"My&#8203;::a"\} = sub \{\}'
Subroutine a redefined at \-e line 1\.

The first two are fine\, the third one loses the package name in report\. Why the second 
one liner reports the whole things as it's used\, whereas the third one loses the 
package name?

This is a problem if functions from different packages get redefined \(e\.g\. via do\(\)  
or else\) and it would be nice if the third case will not lose the package name\.

Also I see no reason why the first one liner shouldn't warn with&#8203;:

&nbsp;  Subroutine main&#8203;::a redefined at \-e line 1\.

to be always consistent\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.7.3:

Configured by stas at Mon Apr  8 00:55:30 SGT 2002.

Summary of my perl5 (revision 5.0 version 7 subversion 3 patch 15780) configuration:
  Platform:
    osname=linux, osvers=2.4.17-16mdksmp, archname=i686-linux-thread-multi
    uname='linux hope.stason.org 2.4.17-16mdksmp #1 smp fri feb 8 14:27:48 cet 2002 i686 unknown '
    config_args='-des -Dprefix=/home/stas/perl/ithread -Dusethreads -Doptimize=-g -Duseshrplib -Dusedevel'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='2.96 20000731 (Mandrake Linux 8.2 2.96-0.74mdk)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.4.so, so=so, useshrplib=true, libperl=libperl.so
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/home/stas/perl/ithread/lib/5.7.3/i686-linux-thread-multi/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL15771


@INC for perl v5.7.3:
    /home/stas/perl/ithread/lib/5.7.3/i686-linux-thread-multi
    /home/stas/perl/ithread/lib/5.7.3
    /home/stas/perl/ithread/lib/site_perl/5.7.3/i686-linux-thread-multi
    /home/stas/perl/ithread/lib/site_perl/5.7.3
    /home/stas/perl/ithread/lib/site_perl
    .


Environment for perl v5.7.3:
    HOME=/home/stas
    LANG=en
    LANGUAGE=en_US:en
    LC_COLLATE=en_US
    LC_CTYPE=en_US
    LC_MESSAGES=en_US
    LC_MONETARY=en_US
    LC_NUMERIC=en_US
    LC_TIME=en_US
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr//bin:/bin:/usr/bin:.:/usr/local/bin:/home/stas/bin:/home/stas/bin:/usr/local/bin:/usr/X11R6/bin:/usr/java/j2re1.4.0/bin/:/home/stas/src/insure/bin.linux2
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh


```
</details>
